drivers/flash/nrf_qspi_nor.c

```diff
@@ -245,72 +239,99 @@ static inline int qspi_get_zephyr_ret_code(nrfx_err_t res)
 {
+#ifdef CONFIG_MULTITHREADING
 	struct qspi_nor_data *dev_data = dev->data;
 
-	pm_device_busy_set(dev);
-
-#ifdef CONFIG_MULTITHREADING
 	k_sem_take(&dev_data->sem, K_FOREVER);
-#else /* CONFIG_MULTITHREADING */
-	ARG_UNUSED(dev_data);
-#endif /* CONFIG_MULTITHREADING */
-
-	/*
-	 * Change the base clock divider only for the time the driver is locked
-	 * to perform a QSPI operation, otherwise the power consumption would be
-	 * increased also when the QSPI peripheral is idle.
-	 * When XIP is enabled, there is nothing to do here as the changed
-	 * divider is kept all the time.
-	 */
-#if defined(CONFIG_SOC_SERIES_NRF53X)
-	if (!dev_data->xip_enabled) {
-		nrf_clock_hfclk192m_div_set(NRF_CLOCK, BASE_CLOCK_DIV);
-	}
 #endif
 }
 
 static inline void qspi_unlock(const struct device *dev)
 {
+#ifdef CONFIG_MULTITHREADING
 	struct qspi_nor_data *dev_data = dev->data;
 
-#if defined(CONFIG_SOC_SERIES_NRF53X)
-	/* Restore the default base clock divider to reduce power consumption.
-	 * Unless XIP is enabled, then the changed divider needs to be kept.
-	 */
-	if (!dev_data->xip_enabled) {
-		nrf_clock_hfclk192m_div_set(NRF_CLOCK, NRF_CLOCK_HFCLK_DIV_4);
-	}
+	k_sem_give(&dev_data->sem);
 #endif
+}
 
-#ifdef CONFIG_MULTITHREADING
-	k_sem_give(&dev_data->sem);
-#else
-	ARG_UNUSED(dev_data);
+static inline void qspi_clock_div_change(void)
+{
+#ifdef CONFIG_SOC_SERIES_NRF53X
+	/* Make sure the base clock divider is changed accordingly
+	 * before a QSPI transfer is performed.
+	 */
+	nrf_clock_hfclk192m_div_set(NRF_CLOCK, BASE_CLOCK_DIV);
 #endif
+}
 
-	pm_device_busy_clear(dev);
+static inline void qspi_clock_div_restore(void)
+{
+#ifdef CONFIG_SOC_SERIES_NRF53X
+	/* Restore the default base clock divider to reduce power
+	 * consumption when the QSPI peripheral is idle.
+	 */
+	nrf_clock_hfclk192m_div_set(NRF_CLOCK, NRF_CLOCK_HFCLK_DIV_4);
+#endif
 }
 
-static inline void qspi_trans_lock(const struct device *dev)
+static void qspi_acquire(const struct device *dev)
 {
-#ifdef CONFIG_MULTITHREADING
 	struct qspi_nor_data *dev_data = dev->data;
 
-	k_sem_take(&dev_data->trans, K_FOREVER);
-#else /* CONFIG_MULTITHREADING */
-	ARG_UNUSED(dev);
-#endif /* CONFIG_MULTITHREADING */
+#if defined(CONFIG_PM_DEVICE_RUNTIME)
+	int rc = pm_device_runtime_get(dev);
+
+	if (rc < 0) {
+		LOG_ERR("pm_device_runtime_get failed: %d", rc);
+	}
+#elif defined(CONFIG_MULTITHREADING)
+	/* In multithreading, the driver can call qspi_acquire more than once
+	 * before calling qspi_release. Keeping count, so QSPI is deactivated
+	 * only at the last call (count == 0).
+	 */
+	k_sem_give(&dev_data->count);
+#endif
+
+	qspi_lock(dev);
+
+	if (!dev_data->xip_enabled) {
+		qspi_clock_div_change();
+
+		pm_device_busy_set(dev);
+	}
 }
 
-static inline void qspi_trans_unlock(const struct device *dev)
+static void qspi_release(const struct device *dev)
 {
-#ifdef CONFIG_MULTITHREADING
 	struct qspi_nor_data *dev_data = dev->data;
+	bool deactivate = true;
```

Comment:
You could probably just assign `deactive = !IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)`

Reply:
Not going to insist on change, marking as resolved.